rfd/0191-workload-id-config-ux.md

```diff
@@ -0,0 +1,730 @@
+
+The attribute key-value set can be represented as a map, with the key being a
+string and the value being a string.
+
+The attribute key-value set will be encoded as JSON into the X509 certificate
+using a new extension (1.3.9999.2.21).
+
+Whilst out of scope of this RFD, the encoding of these values into the X509
+certificate would enable the inclusion of join metadata within audit logs for
+actions taken using Machine ID, allowing specific actions to be directly linked
+to a specific join without correlating the Bot Instance ID from the audit event
+with a `bot.join` event.
+
+### New SVID Issuance RPC
+
+```proto
+// WorkloadIdentityService provides the signing of workload identity documents.
+service WorkloadIdentityService {
+  rpc IssueWorkloadIdentity(IssueWorkloadIdentityRequest) returns (IssueWorkloadIdentityResponse) {}
+}
+
+message WorkloadIdentityNameSelector {
+  string name = 1;
+}
+
+message WorkloadIdentityLabelSelector {
+  map<string, string> labels = 1;
+}
+
+message WorkloadAttestationAttributes {
+  map<string, string> attributes = 2;
+}
+
+message IssueWorkloadIdentityRequest {
+  // Selector.
+  oneof selector {
+    WorkloadIdentityNameSelector name = 1;
+    WorkloadIdentityLabelSelector labels = 2;
+  }
+  WorkloadAttestationAttributes workload_attributes = 3;
+}
+
+message IssueWorkloadIdentityResponse {
+
+}
+```
+
+When a specific WorkloadIdentity is specified by-name by the client:
+
+1. Fetch WorkloadIdentity resource by name.
+2. Compare `workload_identity_labels` of the client roleset against
+  `metadata.labels` of the WorkloadIdentity. Reject if no access.
+3. Perform rules evaluation of `spec.rules.deny` against the attributes of the
+  client. Reject if no access.
+4. If `spec.rules.allow` is non-zero, complete rules evaluation against the
+  attributes of the client. Reject if no access.
+5. Evaluate templated fields, if attribute included in template missing from
+  attributes of the client, reject.
+6. Issue workload identity credential, emit audit log, and return credential to
+  client.
+
+Where the client has provided label selectors:
+
+1. Fetch all WorkloadIdentity resources.
+2. For each WorkloadIdentity, compare client's requested labels against 
+  `metadata.labels`, drop if no match.
+3. For each remaining WorkloadIdentity, compare `workload_identity_labels` of
+  client's roleset against `metadata.labels`, drop if no match.
+4. For each remaining WorkloadIdentity:
+4a. Perform rules evaluation of `spec.rules.deny` against the attributes of the
+   client. Drop if no access.
+4b. If `spec.rules.allow` is non-zero, complete rules evaluation against the
+  attributes of the client. Drop if no access. 
+4c. Evaluate templated fields, if attribute included in template missing from
+  attributes of the client, drop. 
+4d. Issue workload identity credential, emit audit log.
+5. Return issued credentials to client. 
+
+If, after evaluation of labels and rules, more than 10 WorkloadIdentity
```

Comment:
I feel like some sort of limit here is beneficial, but I'm unsure if "10" is the reasonable value. Also, unsure if we should consider this limit before the rules are processed - e.g just after label matching has completed.

Reply:
I could see the limit being problematic if we're nudging users toward relying mostly on this new allow/deny system over Teleport's RBAC - at least depending if this is after label matching (and `role` rule processing) or after `workload_identity` rule processing.

What part of the pipeline do we expect to be the most expensive? The rule processing, or the SPIFFE issuance?

Reply:
I think the SPIFFE issuance is likely to be considered the "expensive" bit, unless we're talking about processing rules for hundreds or thousands of WorkloadIdentity resources - although that could be fixed by some kind of specialised cache down the line.

My main thought as well is that issuing a single workload a large number of identities is probably indicative of a problem. In 99% of use-cases, a workload will expect a single identity.

Reply:
It's always easier to increase limits on customers than reduce them. I don't know if 10 is the reasonable starting place but if you can make it easy to change then keep it low and we can bump it as needed. Is that possible? Or would changing it require a new version or something?

Reply:
952b87991b